internal/controller/nonadmindownloadrequest_controller.go

```diff
@@ -0,0 +1,299 @@
+			// find associated velero downloadrequest and delete that first
+			return reconcile.Result{}, r.deleteExpiredDownloadRequest(ctx, &veleroDR, req)
+		}
+		// if request is not expired, and has downloadUrl, requeue when expired to cleanup.
+		if req.Status.VeleroDownloadRequest.Status.DownloadURL != constant.EmptyString {
+			return reconcile.Result{RequeueAfter: time.Until(req.Status.VeleroDownloadRequest.Status.Expiration.Time)}, nil
+		}
+	}
+	// try get veleroDR if exists, then update status
+	if err := r.Get(ctx, types.NamespacedName{Namespace: veleroDR.Namespace, Name: veleroDR.Name}, &veleroDR); err == nil {
+		return reconcile.Result{}, r.updateNADRWithURL(ctx, &veleroDR, req)
+	} else if !apierrors.IsNotFound(err) {
+		// some other errors, requeue to retry get
+		return reconcile.Result{}, err
+	}
+	return reconcile.Result{}, r.processDownloadRequest(ctx, &veleroDR, req)
+}
+
+// Delete expired download request
+// find associated velero downloadrequest and delete that first
+func (r *NonAdminDownloadRequestReconciler) deleteExpiredDownloadRequest(ctx context.Context, veleroDR *velerov1.DownloadRequest, req *nacv1alpha1.NonAdminDownloadRequest) error {
+	logger := log.FromContext(ctx)
+	logger.V(1).Info("Deleting expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), req.Namespace)
+	if err := r.Delete(ctx, veleroDR); err != nil {
+		if !apierrors.IsNotFound(err) {
+			logger.V(1).Info("Failed to delete expired NonAdminDownloadRequest associated velero download request", req.VeleroDownloadRequestName(), err)
+			// other errors, requeue to retry delete
+			return err
+		}
+	}
+	logger.V(1).Info("Deleting expired NonAdminDownloadRequest", req.Name, req.Namespace)
+	if err := r.Delete(ctx, req); err != nil {
+		if !apierrors.IsNotFound(err) {
+			logger.V(1).Info("Failed to delete expired NonAdminDownloadRequest", req.Name, err)
+			// other errors, requeue to retry delete
+			return err
+		}
+		// not found, stop requeue
+	}
+	return nil
+}
+
+// if url is available and not expired, then set status to completed
+func (r *NonAdminDownloadRequestReconciler) updateNADRWithURL(ctx context.Context, veleroDR *velerov1.DownloadRequest, req *nacv1alpha1.NonAdminDownloadRequest) error {
+	logger := log.FromContext(ctx)
+	if veleroDR.Status.DownloadURL != constant.EmptyString &&
+		veleroDR.Status.Phase == velerov1.DownloadRequestPhaseProcessed &&
+		!veleroDR.Status.Expiration.Before(&metav1.Time{Time: time.Now()}) {
+		// copy req, prepare to patch
+		prePatch := req.DeepCopy()
+		// copy status to NADR from VDR
+		req.Status.VeleroDownloadRequest.Status = &veleroDR.Status
+		req.Status.Phase = nacv1alpha1.NonAdminPhaseCreated
+		// veleroDR is processed, update NADR status
+		// clear conditions
+		req.Status.Conditions = []metav1.Condition{}
```

Comment:
instead of just clearing the conditions, should we also add a `sucessfullyProcessed` condition or something here ?

Reply:
I see condition = blockers. So clearing them = done.  Wdyt.

Reply:
Hmm for other objects we aren't doing this. Lets follow the same pattern here.

Reply:
Also, conditions don't usually/always mean blockers, right ?

Reply:
not always. this follows dpa. sure will change to match other nac